File: app/src/main/java/com/vladislav/workoutassistant/ui/workouts/WorkoutDetailsFragment.java
package com.vladislav.workoutassistant.ui.workouts;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.vladislav.workoutassistant.R;
import com.vladislav.workoutassistant.ui.main.GeneralFragment;
import com.vladislav.workoutassistant.ui.main.interfaces.OnItemClickCallback;
import com.vladislav.workoutassistant.data.models.WorkoutExercise;
import com.vladislav.workoutassistant.data.models.WorkoutProgram;
import com.vladislav.workoutassistant.data.models.WorkoutSet;
import com.vladislav.workoutassistant.ui.exercises.ExercisesFragment;
import com.vladislav.workoutassistant.ui.workouts.adapters.SetAndExerciseAdapter;
import com.vladislav.workoutassistant.ui.main.components.DividerItemDecoration;
import com.vladislav.workoutassistant.ui.main.components.SimpleItemTouchHelperCallback;
import com.vladislav.workoutassistant.ui.workouts.viewmodels.WorkoutViewModel;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

public class WorkoutDetailsFragment extends GeneralFragment implements SimpleItemTouchHelperCallback.OnStartDragListener {

    private static final String WORKOUT_ID_ARG = "workout_id_arg";
    private static final String TITLE_ARG = "title_arg";

    private RecyclerView mProgramRecyclerView;
    private WorkoutViewModel mWorkoutViewModel;
    private SetAndExerciseAdapter mAdapter;
    private ItemTouchHelper mItemTouchHelper;

    private OnItemClickCallback mCallback = new OnItemClickCallback() {
        @Override
        public void onClick(int id, String name) {
            //TODO: some actions
        }
    };

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_workout_details, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        getMFragmentListener().updateToolbarTitle(getArguments().getString(TITLE_ARG));
        setHasOptionsMenu(true);

        mProgramRecyclerView = view.findViewById(R.id.recycler_view);

        mWorkoutViewModel = ViewModelProviders.of(this).get(WorkoutViewModel.class);
        mAdapter = new SetAndExerciseAdapter(mCallback, this, mWorkoutViewModel.getMuscleGroups());
        mWorkoutViewModel.getSetsAndExercisesList().observe(this, new Observer<List<Object>>() {
            @Override
            public void onChanged(List<Object> list) {
                if (list != null) {
                    mAdapter.setList(list);
                    mAdapter.notifyDataSetChanged();
                }
            }
        });
        mWorkoutViewModel.init(getArguments().getInt(WORKOUT_ID_ARG));
        mProgramRecyclerView.setAdapter(mAdapter);
        mProgramRecyclerView.addItemDecoration(new DividerItemDecoration(getActivity(), R.drawable.divider, 20));

        mItemTouchHelper = new ItemTouchHelper(new SimpleItemTouchHelperCallback(mAdapter));
        mItemTouchHelper.attachToRecyclerView(mProgramRecyclerView);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.workouts_actions, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.add_exercises_action:
                getMFragmentListener().addFragmentOnTop(ExercisesFragment.Companion.newInstance(true));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onStartDrag(RecyclerView.ViewHolder viewHolder) {
        mItemTouchHelper.startDrag(viewHolder);
    }

    public static WorkoutDetailsFragment newInstance(int id, String title) {
        Bundle args = new Bundle();
        args.putInt(WORKOUT_ID_ARG, id);
        args.putString(TITLE_ARG, title);

        WorkoutDetailsFragment fragment = new WorkoutDetailsFragment();
        fragment.setArguments(args);

        return fragment;
    }
}

File: app/src/main/java/com/vladislav/workoutassistant/ui/workouts/WorkoutDetailsFragment.kt
package com.vladislav.workoutassistant.ui.workouts

import android.os.Bundle
import android.view.*
import android.widget.Toast
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.RecyclerView
import com.vladislav.workoutassistant.R
import com.vladislav.workoutassistant.ui.exercises.ExercisesFragment
import com.vladislav.workoutassistant.ui.main.GeneralFragment
import com.vladislav.workoutassistant.ui.main.components.DividerItemDecoration
import com.vladislav.workoutassistant.ui.main.interfaces.OnItemClickCallback
import com.vladislav.workoutassistant.ui.workouts.adapters.WorkoutExerciseAdapter
import com.vladislav.workoutassistant.ui.workouts.viewmodels.WorkoutViewModel

class WorkoutDetailsFragment : GeneralFragment() {

    private val mWorkoutViewModel: WorkoutViewModel by lazy {
        ViewModelProviders.of(this).get(WorkoutViewModel::class.java)
    }
    private val mCallback = object : OnItemClickCallback {
        override fun onClick(id: Int, name: String) {
            Toast.makeText(activity, "$name $id", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_workout_details, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        mFragmentListener!!.updateToolbarTitle(arguments!!.getString(TITLE_ARG)!!)
        setHasOptionsMenu(true)

        val mAdapter = WorkoutExerciseAdapter(activity!!.application, mCallback)

        val recyclerView = view.findViewById<RecyclerView>(R.id.recycler_view)
        recyclerView.adapter = mAdapter
        recyclerView.addItemDecoration(DividerItemDecoration(activity, R.drawable.divider, 20))

        mWorkoutViewModel.workoutInfo.observe(this, Observer { info ->
            mAdapter.updateContent(info.sets)
        })
        mWorkoutViewModel.loadWorkoutInfoById(arguments!!.getInt(WORKOUT_ID_ARG))
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.workouts_actions, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.add_exercises_action -> {
                mFragmentListener?.addFragmentOnTop(ExercisesFragment.newInstance(true))
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }



    companion object {

        private const val WORKOUT_ID_ARG = "workout_id_arg"
        private const val TITLE_ARG = "title_arg"

        fun newInstance(id: Int, title: String): WorkoutDetailsFragment {
            val args = Bundle()
            args.putInt(WORKOUT_ID_ARG, id)
            args.putString(TITLE_ARG, title)

            val fragment = WorkoutDetailsFragment()
            fragment.arguments = args

            return fragment
        }
    }
}